contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,131 @@
+---
+title: Why and how we moved exports to Temporal
+date: 2023-08-24
+author: ["tomas-farias", "ian-vanagas"]
+showTitle: true
+rootpage: /blog
+sidebar: Blog
+hideAnchor: true
+featuredImage: ../images/blog/posthog-engineering-blog.png
+featuredImageType: full
+category: Engineering
+---
+
+Letting users maintain ownership of their data is a key value of PostHog. Although we aspire to become the single source of truth for all your data needs, we realize we don’t fulfill everything everyone wants to do (yet). Also, we know that even if we were the single source of truth, users still want to use their data on other platforms.
+
+Export apps provided the ability to maintain ownership of all of your data in PostHog and move it where you like. Unfortunately, it didn’t scale as well as we wanted. Luckily for you, we recently launched our [newly upgraded exports system](/docs/cdp/batch-exports), built using Temporal. This post covers why and how we made the change to Temporal.
+
+## The trouble with export apps
+
+Although exports worked well most of the time, they were causing trouble for some customers. They were failing, erroring, duplicating, and dropping data. Fixing these issues was taking up a significant amount of the pipeline team’s time. 
+
+Although stateless plugins (like GeoIP) worked well with our [plugin service structure](/blog/how-we-built-an-app-server), **stateful** plugins (like export apps) didn’t. This is because they depended on external resources that can degrade or fail. Dealing with this required timeout and retry logic that was difficult to build into the existing structure.
+
+Export apps scaled as far as possible. They worked well for smaller users, but as the export requirements grew, it started to show issues. Chiefly among them was that they were:
+
+### 1. Unreliable
+
+The export manager had no way of reliably detecting whether an export had failed or was simply taking too long, and it would eagerly retry after a timeout.
+
+This caused multiple copies of the same export job with the same state to run at the same time. This created a race condition where whenever a new copy started, it would reset the progress to zero, and other copies would resume from the beginning again. This caused duplication in the exported user data, high infrastructure load, and database locks and conflicts.
```

Comment:
```suggestion
First, it generated multiple copies of the same export job with the same state, running at the same time. This created a race condition where, whenever a new copy started, it would reset the progress to zero, and other copies would resume from the beginning again. The resulted in duplication in the exported user data, high infrastructure load, and database locks and conflicts.
```